src/lib/common/payload_uart.cpp

```diff
@@ -28,6 +28,97 @@ static bool _useLineBuffer = true;
+static bool _writeDuringTransaction = false;
+// Optional transaction start function
+static HardwareControlFunc _preTxFunction = nullptr;
+// Optional transaction end function, will be called after final write of transaction
+static HardwareControlFunc _postTxFunction = nullptr;
+// Semaphore for passing event to postTxCallbackTask
+static SemaphoreHandle_t _postTxSemaphore = NULL;
+// Flag to prevent double triggers of the _postTxFunction callback
+static volatile bool _postTxFunctionPending = false;
+
+// Forward declaration of the post-transmission callback function
+static void pluartPostTxCallback(SerialHandle_t *handle);
+
+// Set transaction hardware control functions
+void enableTransactions(HardwareControlFunc preTxFunc, HardwareControlFunc postTxFunc) {
+  // Assign the transaction start and end callbacks
+  _preTxFunction = preTxFunc;
+  _postTxFunction = postTxFunc;
+  // Assign the post-transmission callback
+  uart_handle.postTxCb = pluartPostTxCallback;
+}
+
+// Start a Tx transaction
+void startTransaction() {
+  _transactionInProgress = true;
+  _writeDuringTransaction = false;
+  if (_preTxFunction) {
+    _preTxFunction(); // External start transaction function. Eg, enable Tx for Half-Duplex
+  }
+}
+
+// End a Tx transaction
+// _postTxFunction call will be triggered either from here if no writes are pending,
+//  or by TC interrupt firing on completion of the final write (see lib/common/serial.c).
+//  Use a semaphore and a separate task to trigger and execute this safely.
+void endTransaction() {
+  taskENTER_CRITICAL();
+  if (_transactionInProgress && _postTxFunction) {
+    // ensure _postTxFunction is called if we start and end transaction without writing
+    // ensure _postTxFunction is called if write completes and fires TC interrupt before we've
+    //  marked the transaction as complete.
+    _transactionInProgress = false;
+    if (!_writeDuringTransaction || xStreamBufferIsEmpty(uart_handle.txStreamBuffer)) {
+      // Transmission is complete, ensure callback triggers
+      if (!_postTxFunctionPending) {
+        _postTxFunctionPending = true;
+        taskEXIT_CRITICAL();
+        xSemaphoreGive(_postTxSemaphore);
+      } else {
+        taskEXIT_CRITICAL();
+      }
+    } else {
+      // Serial writes are in progress, and we will wait for them to complete.
+      // uart_handle.postTxCb (pluartPostTxCallback) will trigger call to _postTxFunction on
+      // completion of write(). No further action needed here
+      taskEXIT_CRITICAL();
+    }
+  } else {
+    _transactionInProgress = false;
+    taskEXIT_CRITICAL();
+  }
+}
+
+// Post-transmission callback function, attached as uart_handle.postTxCb and called
+//  from ISR when the TC interrupt fires [see common/serial.c::serialGenericUartIRQHandler].
+static void pluartPostTxCallback(SerialHandle_t *handle) {
+  (void)handle; // Mark 'handle' as intentionally unused
+
+  BaseType_t xHigherPriorityTaskWoken = pdFALSE;
```

Comment:
@towynlin @victorsowa12 @matt001k  FYI - I don't fully grok the safety handling for ENTER / EXIT_CRITICAL_FROM_ISR, which includes detection and yielding if higher priority preemption, preserving saved interrupt context.

This was all suggested by ChatGPT. 😬

Reply:
This is an appropriate use case for entering/exiting sections critical sections within ISR context (prevents lower priority interrupts from pre-empting this service routine). The yielding will essentially allow the `postTxCBTask` task to pre-empt other tasks of lower priority which is necessary due to the semaphore giving here. The answer to this question provides a nice graphical look at what not running this would look like: https://stackoverflow.com/questions/58631246/freertos-why-to-call-taskyield-from-isr-method-within-the-isrhandler